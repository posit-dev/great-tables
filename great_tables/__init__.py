# Set version ----
from importlib_metadata import version as _v

__version__ = _v("great_tables")

del _v

# Main gt imports ----

from .gt import GT
from . import vals, loc, style
from ._styles import FromColumn as from_column
from ._helpers import (
    letters,
    LETTERS,
    px,
    pct,
    md,
    html,
    google_font,
    random_id,
    system_fonts,
    define_units,
    nanoplot_options,
)


__all__ = (
    "GT",
    "exibble",
    "letters",
    "LETTERS",
    "px",
    "pct",
    "md",
    "html",
    "google_font",
    "system_fonts",
    "define_units",
    "nanoplot_options",
    "random_id",
    "from_column",
    "vals",
    "loc",
    "style",
)


def __getattr__(k: str):
    # exibble dataset available on top-level module, but is a pandas DataFrame.
    # Since pandas is an optional dependency, we import exibble dynamically.
    if k == "exibble":
        from great_tables.data import exibble

        return exibble

    # allow the data submodule to be accessed directly, as if it were a top-level import
    if k == "data":
        import great_tables.data

        return great_tables.data

    raise AttributeError(f"cannot get attribute {k} from great_tables ({__file__})")
