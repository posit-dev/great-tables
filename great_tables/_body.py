from __future__ import annotations

from typing import TYPE_CHECKING

from ._tbl_data import copy_data

if TYPE_CHECKING:
    from ._gt_data import Body


def body_reassemble(body: Body) -> Body:
    # Note that this used to order the body based on groupings, but now that occurs in the
    # renderer itself.
    return body.__class__(copy_data(body.body))
