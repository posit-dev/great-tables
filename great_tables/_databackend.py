from __future__ import annotations

import importlib
from abc import ABCMeta
from types import ModuleType


def _load_class(mod_name: str, cls_name: str) -> type:
    # follow the path of mods if it is one
    listified_mod_path: list[str] = cls_name.split(".")

    sub_module_path: list[str] = [mod_name] + listified_mod_path[:-1]

    cur_mod: ModuleType = importlib.import_module(sub_module_path[0])
    _cls_name: str = listified_mod_path[-1]

    if len(sub_module_path) == 1:  # no submodules, just the module itself
        return getattr(cur_mod, _cls_name)

    # now, we must increment the module paths to import the final class
    remaining_mods: list[str] = sub_module_path[1:]
    for mod in remaining_mods:
        cur_mod: ModuleType = getattr(cur_mod, mod)

    return getattr(cur_mod, _cls_name)


class _AbstractBackendMeta(ABCMeta):
    def register_backend(cls, mod_name: str, cls_name: str):
        cls._backends.append((mod_name, cls_name))
        cls._abc_caches_clear()


class AbstractBackend(metaclass=_AbstractBackendMeta):
    @classmethod
    def __init_subclass__(cls):
        if not hasattr(cls, "_backends"):
            cls._backends = []

    @classmethod
    def __subclasshook__(cls, subclass):
        for mod_name, cls_name in cls._backends:
            try:
                parent_candidate = _load_class(mod_name, cls_name)
                if issubclass(subclass, parent_candidate):
                    return True
            except (ModuleNotFoundError, ImportError):
                continue

        return NotImplemented
