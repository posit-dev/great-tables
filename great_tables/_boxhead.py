from __future__ import annotations

from typing import TYPE_CHECKING

from ._locations import resolve_cols_c
from ._utils import _assert_list_is_subset
from ._tbl_data import SelectExpr
from ._text import BaseText

if TYPE_CHECKING:
    from ._types import GTSelf


def cols_label(
    self: GTSelf, cases: dict[str, str | BaseText] | None = None, **kwargs: str | BaseText
) -> GTSelf:
    """
    Relabel one or more columns.

    There are three important pieces to labelling:

    * Each argument has the form: {name in data} = {new label}.
    * Multiple columns may be given the same label.
    * Labels may use curly braces to apply special formatting, called unit notation.
      For example, "area ({{ft^2}})" would appear as "area (ftÂ²)".

    See [`define_units()`](`great_tables.define_units`) for details on unit notation.

    Parameters
    ----------
    cases
        A dictionary where the keys are column names and the values are the labels. Labels may use
        [`md()`](`great_tables.md`) or [`html()`](`great_tables.html`) helpers for formatting.

    **kwargs
        Keyword arguments to specify column labels. Each keyword corresponds to a column name, with
        its value indicating the new label.

    Returns
    -------
    GT
        The GT object is returned. This is the same object that the method is called on so that we
        can facilitate method chaining.

    Notes
    -----
    GT always selects columns using their name in the underlying data. This means that a column's
    label is purely for final presentation.

    Examples
    --------

    The example below relabels columns from the `countrypops` data to start with uppercase.

    ```{python}
    from great_tables import GT
    from great_tables.data import countrypops

    countrypops_mini = countrypops.loc[countrypops["country_name"] == "Uganda"][
        ["country_name", "year", "population"]
    ].tail(5)

    (
        GT(countrypops_mini)
        .cols_label(
            country_name="Country Name",
            year="Year",
            population="Population"
        )
    )
    ```

    Note that we supplied the name of the column as the key, and the new label as the value.

    We can also use Markdown formatting for the column labels. In this example, we'll use
    `md("*Population*")` to make the label italicized.

    ```{python}
    from great_tables import GT, md
    from great_tables.data import countrypops

    (
        GT(countrypops_mini)
        .cols_label(
            country_name="Name",
            year="Year",
            population=md("*Population*")
        )
    )
    ```

    We can also use unit notation to format the column labels. In this example, we'll use
    `{{cm^3 molecules^-1 s^-1}}` for part of the label for the `OH_k298` column.

    ```{python}
    from great_tables import GT
    from great_tables.data import reactions
    import polars as pl

    reactions_mini = (
        pl.from_pandas(reactions)
        .filter(pl.col("cmpd_type") == "mercaptan")
        .select(["cmpd_name", "OH_k298"])
    )

    (
        GT(reactions_mini)
        .fmt_scientific("OH_k298")
        .sub_missing()
        .cols_label(
            cmpd_name="Compound Name",
            OH_k298="OH, {{cm^3 molecules^-1 s^-1}}",
        )
    )
    ```
    """
    from great_tables._helpers import UnitStr

    cases = cases if cases is not None else {}
    new_cases = cases | kwargs

    # If nothing is provided, return `data` unchanged
    if len(new_cases) == 0:
        return self

    # Get the full list of column names for the data
    column_names = self._boxhead._get_columns()
    mod_columns = list(new_cases.keys())

    # Stop function if any of the column names specified are not in `cols_labels`
    # msg: "All column names provided must exist in the input `.data` table."
    _assert_list_is_subset(mod_columns, set_list=column_names)

    # Handle units syntax in labels (e.g., "Density ({{ppl / mi^2}})")
    new_kwargs: dict[str, UnitStr | str | BaseText] = {}

    for k, v in new_cases.items():
        if isinstance(v, str):
            unitstr_v = UnitStr.from_str(v)

            if len(unitstr_v.units_str) == 1 and isinstance(unitstr_v.units_str[0], str):
                new_kwargs[k] = unitstr_v.units_str[0]
            else:
                new_kwargs[k] = unitstr_v

        elif isinstance(v, BaseText):
            new_kwargs[k] = v

        else:
            raise ValueError(
                "Column labels must be strings or BaseText objects. Use `md()` or `html()` for formatting."
            )

    boxhead = self._boxhead._set_column_labels(new_kwargs)

    return self._replace(_boxhead=boxhead)


def cols_align(self: GTSelf, align: str = "left", columns: SelectExpr = None) -> GTSelf:
    """
    Set the alignment of one or more columns.

    The `cols_align()` method sets the alignment of one or more columns. The `align` argument
    can be set to one of `"left"`, `"center"`, or `"right"` and the `columns` argument can be
    used to specify which columns to apply the alignment to. If `columns` is not specified, the
    alignment is applied to all columns.

    Parameters
    ----------
    align
        The alignment to apply. Must be one of `"left"`, `"center"`, or `"right"`.
    columns
        The columns to target. Can either be a single column name or a series of column names
        provided in a list. If `None`, the alignment is applied to all columns.

    Returns
    -------
    GT
        The GT object is returned. This is the same object that the method is called on so that we
        can facilitate method chaining.

    Examples
    --------
    Let's use the `countrypops` to create a small table. We can change the alignment of the
    `population` column with `cols_align()`. In this example, the column label and body cells of
    `population` will be aligned to the left.

    ```{python}
    from great_tables import GT
    from great_tables.data import countrypops

    countrypops_mini = countrypops.loc[countrypops["country_name"] == "San Marino"][
        ["country_name", "year", "population"]
    ].tail(5)

    (
        GT(countrypops_mini, rowname_col="year", groupname_col="country_name")
        .cols_align(align="left", columns="population")
    )
    ```

    """

    # Throw if `align` is not one of the three allowed values
    if align not in ["left", "center", "right"]:
        raise ValueError("Align must be one of 'left', 'center', or 'right'.")

    # Get the full list of column names for the data
    column_names = self._boxhead._get_columns()

    # Upgrade `columns` to a list if `columns` is a string and not None
    if isinstance(columns, str):
        columns = [columns]
        _assert_list_is_subset(columns, set_list=column_names)
    elif columns is None:
        columns = column_names

    sel_cols = resolve_cols_c(data=self, expr=columns)

    # Set the alignment for each column
    return self._replace(_boxhead=self._boxhead._set_column_aligns(sel_cols, align=align))
