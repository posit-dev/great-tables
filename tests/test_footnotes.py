import polars as pl
import re
from great_tables import GT, loc


def _create_test_data():
    # Create DataFrame with potential for stub and two row groups
    return pl.DataFrame(
        {
            "group": ["A", "A", "A", "B", "B", "B"],
            "row_id": ["r1", "r2", "r3", "r4", "r5", "r6"],
            "col1": [10, 20, 30, 40, 50, 60],
            "col2": [100, 200, 300, 400, 500, 600],
            "col3": [1000, 2000, 3000, 4000, 5000, 6000],
        }
    )


def _create_base_gt():
    df = _create_test_data()
    return (
        GT(df, rowname_col="row_id", groupname_col="group")
        .tab_header(title="Test Title", subtitle="Test Subtitle")
        .tab_spanner(label="Spanner", columns=["col1", "col2"])
    )


def test_tab_footnote_basic():
    # Test basic footnote creation and HTML rendering
    gt_table = _create_base_gt().tab_footnote(
        footnote="Test footnote", locations=loc.body(columns="col1", rows=[0])
    )

    html = gt_table._render_as_html()

    # Check that footnote appears in footer
    assert "Test footnote" in html
    # Check that footnote mark appears in cell
    assert re.search(r"10<span[^>]*>1</span>", html)


def test_tab_footnote_numeric_marks():
    # Test numeric footnote marks (default type of marks)
    gt_table = (
        _create_base_gt()
        .tab_footnote(footnote="First note", locations=loc.body(columns="col1", rows=[0]))
        .tab_footnote(footnote="Second note", locations=loc.body(columns="col2", rows=[1]))
        .tab_footnote(footnote="Third note", locations=loc.body(columns="col3", rows=[2]))
    )

    html = gt_table._render_as_html()

    # Check that marks appear in the correct order
    assert re.search(r"10<span[^>]*>1</span>", html)  # First cell
    assert re.search(r"200<span[^>]*>2</span>", html)  # Second cell
    assert re.search(r"3000<span[^>]*>3</span>", html)  # Third cell


def test_tab_footnote_mark_coalescing():
    # Test that multiple footnotes on same location show up as comma-separated marks
    gt_table = (
        _create_base_gt()
        .tab_footnote(footnote="First note", locations=loc.body(columns="col1", rows=[0]))
        .tab_footnote(footnote="Second note", locations=loc.body(columns="col1", rows=[0]))
        .tab_footnote(footnote="Third note", locations=loc.body(columns="col2", rows=[1]))
    )

    html = gt_table._render_as_html()

    # First cell should have coalesced marks "1,2"
    assert re.search(r"10<span[^>]*>1,2</span>", html)
    # Second cell should have single mark "3"
    assert re.search(r"200<span[^>]*>3</span>", html)


def test_tab_footnote_ordering():
    # Test that footnotes are ordered left-to-right, top-to-bottom
    gt_table = (
        _create_base_gt()
        .tab_footnote(footnote="Body note", locations=loc.body(columns="col1", rows=[0]))
        .tab_footnote(footnote="Header note", locations=loc.column_labels(columns="col1"))
        .tab_footnote(footnote="Later body note", locations=loc.body(columns="col2", rows=[1]))
    )

    html = gt_table._render_as_html()

    # Header should get mark 1 (comes before body)
    assert re.search(r">col1<span[^>]*>1</span>", html)
    # First body cell should get mark 2
    assert re.search(r"10<span[^>]*>2</span>", html)
    # Later body cell should get mark 3
    assert re.search(r"200<span[^>]*>3</span>", html)


def test_tab_footnote_all_locations():
    # Test that footnotes can be placed in all major locations
    gt_table = (
        _create_base_gt()
        .tab_footnote(footnote="Title note", locations=loc.title())
        .tab_footnote(footnote="Subtitle note", locations=loc.subtitle())
        .tab_footnote(footnote="Spanner note", locations=loc.spanner_labels(ids=["Spanner"]))
        .tab_footnote(footnote="Column note", locations=loc.column_labels(columns="col1"))
        .tab_footnote(footnote="Body note", locations=loc.body(columns="col1", rows=[0]))
        .tab_footnote(footnote="Stub note", locations=loc.stub(rows=[0]))
        .tab_footnote(footnote="Row group note", locations=loc.row_groups(rows=[0]))
    )

    html = gt_table._render_as_html()

    # All footnotes should appear in footer
    for note in [
        "Title note",
        "Subtitle note",
        "Spanner note",
        "Column note",
        "Body note",
        "Stub note",
        "Row group note",
    ]:
        assert note in html

    # Check that the footnote marks in the title and subtitle appear
    assert re.search(r"Test Title<span[^>]*>1</span>", html)  # Title
    assert re.search(r"Test Subtitle<span[^>]*>2</span>", html)  # Subtitle


def test_tab_footnote_symbol_marks_standard():
    # Test "standard" symbol marks
    gt_table = (
        _create_base_gt()
        .tab_footnote(footnote="First note", locations=loc.body(columns="col1", rows=[0]))
        .tab_footnote(footnote="Second note", locations=loc.body(columns="col2", rows=[1]))
        .tab_footnote(footnote="Third note", locations=loc.body(columns="col3", rows=[2]))
        .tab_footnote(footnote="Fourth note", locations=loc.body(columns="col1", rows=[1]))
        .opt_footnote_marks("standard")
    )

    html = gt_table._render_as_html()

    # Check standard symbols appear in visual reading order
    assert re.search(r"10<span[^>]*>\*</span>", html)
    assert re.search(r"20<span[^>]*>†</span>", html)
    assert re.search(r"200<span[^>]*>‡</span>", html)
    assert re.search(r"3000<span[^>]*>§</span>", html)


def test_tab_footnote_symbol_marks_extended():
    # Test "extended" symbol marks
    gt_table = (
        _create_base_gt()
        .tab_footnote(footnote="Note 1", locations=loc.body(columns="col1", rows=[0]))
        .tab_footnote(footnote="Note 2", locations=loc.body(columns="col2", rows=[0]))
        .tab_footnote(footnote="Note 3", locations=loc.body(columns="col3", rows=[0]))
        .tab_footnote(footnote="Note 4", locations=loc.body(columns="col1", rows=[1]))
        .tab_footnote(footnote="Note 5", locations=loc.body(columns="col2", rows=[1]))
        .tab_footnote(footnote="Note 6", locations=loc.body(columns="col3", rows=[1]))
        .opt_footnote_marks("extended")
    )

    html = gt_table._render_as_html()

    # Check extended symbols appear in reading order (left-to-right, top-to-bottom)
    symbols = ["*", "†", "‡", "§", "‖", "¶"]
    values = [10, 100, 1000, 20, 200, 2000]

    for symbol, value in zip(symbols, values):
        escaped_symbol = re.escape(symbol)
        assert re.search(f"{value}<span[^>]*>{escaped_symbol}</span>", html)


def test_tab_footnote_symbol_marks_letters():
    # Test letter-based marks ("letters")
    gt_table = (
        _create_base_gt()
        .tab_footnote(footnote="Note A", locations=loc.body(columns="col1", rows=[0]))
        .tab_footnote(footnote="Note B", locations=loc.body(columns="col2", rows=[0]))
        .tab_footnote(footnote="Note C", locations=loc.body(columns="col3", rows=[0]))
        .opt_footnote_marks("letters")
    )

    html = gt_table._render_as_html()

    # Check that the letter marks appear
    assert re.search(r"10<span[^>]*>a</span>", html)
    assert re.search(r"100<span[^>]*>b</span>", html)
    assert re.search(r"1000<span[^>]*>c</span>", html)


def test_tab_footnote_symbol_marks_uppercase_letters():
    # Test uppercase letter marks ("LETTERS")
    gt_table = (
        _create_base_gt()
        .tab_footnote(footnote="Note A", locations=loc.body(columns="col1", rows=[0]))
        .tab_footnote(footnote="Note B", locations=loc.body(columns="col2", rows=[0]))
        .tab_footnote(footnote="Note C", locations=loc.body(columns="col3", rows=[0]))
        .opt_footnote_marks("LETTERS")
    )

    html = gt_table._render_as_html()

    # Check that the uppercase letter marks appear
    assert re.search(r"10<span[^>]*>A</span>", html)
    assert re.search(r"100<span[^>]*>B</span>", html)
    assert re.search(r"1000<span[^>]*>C</span>", html)


def test_tab_footnote_custom_symbol_marks():
    # Test custom symbol marks
    custom_marks = ["❶", "❷", "❸", "❹"]  # using circled numbers
    gt_table = (
        _create_base_gt()
        .tab_footnote(footnote="Note 3", locations=loc.body(columns="col3", rows=[0]))
        .tab_footnote(footnote="Note 2", locations=loc.body(columns="col2", rows=[0]))
        .tab_footnote(footnote="Note 1", locations=loc.body(columns="col1", rows=[0]))
        .opt_footnote_marks(custom_marks)
    )

    html = gt_table._render_as_html()

    # Check that the custom marks appear (in the right order)
    assert re.search(r"10<span[^>]*>❶</span>", html)
    assert re.search(r"100<span[^>]*>❷</span>", html)
    assert re.search(r"1000<span[^>]*>❸</span>", html)


def test_tab_footnote_symbol_cycling():
    # Test the symbol cycling feature (when there are more footnotes than symbols)
    gt_table = (
        _create_base_gt()
        .tab_footnote(footnote="Note 1", locations=loc.body(columns="col1", rows=[0]))
        .tab_footnote(footnote="Note 2", locations=loc.body(columns="col2", rows=[0]))
        .tab_footnote(footnote="Note 3", locations=loc.body(columns="col3", rows=[0]))
        .tab_footnote(
            footnote="Note 4", locations=loc.body(columns="col1", rows=[1])
        )  # Should cycle to **
        .tab_footnote(
            footnote="Note 5", locations=loc.body(columns="col2", rows=[1])
        )  # Should cycle to ††
        .opt_footnote_marks("standard")
    )

    html = gt_table._render_as_html()

    # Check the cycling behavior
    assert re.search(r"10<span[^>]*>\*</span>", html)
    assert re.search(r"100<span[^>]*>†</span>", html)
    assert re.search(r"1000<span[^>]*>‡</span>", html)
    assert re.search(r"20<span[^>]*>§</span>", html)
    assert re.search(r"200<span[^>]*>\*\*</span>", html)


def test_tab_footnote_symbol_coalescing():
    # Test symbol mark coalescing with commas
    gt_table = (
        _create_base_gt()
        .tab_footnote(footnote="First note", locations=loc.body(columns="col1", rows=[0]))
        .tab_footnote(footnote="Second note", locations=loc.body(columns="col1", rows=[0]))
        .tab_footnote(footnote="Third note", locations=loc.body(columns="col2", rows=[0]))
        .opt_footnote_marks("standard")
    )

    html = gt_table._render_as_html()

    # The first cell should have a coalesced symbol marks
    assert re.search(r"10<span[^>]*>\*,†</span>", html)
    # The second cell should have a single symbol mark
    assert re.search(r"100<span[^>]*>‡</span>", html)


def test_tab_footnote_multiple_rows():
    # Test a single footnote targeting multiple rows
    gt_table = _create_base_gt().tab_footnote(
        footnote="Multiple rows note", locations=loc.body(columns="col1", rows=[0, 1, 2])
    )

    html = gt_table._render_as_html()

    # All three cells should have the same footnote mark
    assert re.search(r"10<span[^>]*>1</span>", html)
    assert re.search(r"20<span[^>]*>1</span>", html)
    assert re.search(r"30<span[^>]*>1</span>", html)


def test_tab_footnote_multiple_columns():
    # Test footnote targeting multiple columns
    gt_table = _create_base_gt().tab_footnote(
        footnote="Multiple columns note", locations=loc.body(columns=["col1", "col2"], rows=[0])
    )

    html = gt_table._render_as_html()

    # Both cells in the first row should have the same footnote mark
    assert re.search(r"10<span[^>]*>1</span>", html)
    assert re.search(r"100<span[^>]*>1</span>", html)


def test_tab_footnote_footer_rendering():
    # Test that the footnotes section is properly rendered
    gt_table = (
        _create_base_gt()
        .tab_footnote(footnote="First footnote text", locations=loc.body(columns="col1", rows=[0]))
        .tab_footnote(footnote="Second footnote text", locations=loc.body(columns="col2", rows=[1]))
        .opt_footnote_marks("standard")
    )

    html = gt_table._render_as_html()

    # Check footnotes appear in footer with correct marks
    footer_match = re.search(r"<tfoot[^>]*>.*?</tfoot>", html, re.DOTALL)
    assert footer_match is not None
    footer_html = footer_match.group(0)

    assert re.search(r"<span[^>]*>\*</span>\s*First footnote text", footer_html)
    assert re.search(r"<span[^>]*>†</span>\s*Second footnote text", footer_html)


def test_tab_footnote_with_text_object():
    # Test a footnote with the Text object (not using a basic string)
    from great_tables._text import Text

    gt_table = _create_base_gt().tab_footnote(
        footnote=Text("Bold text"), locations=loc.body(columns="col1", rows=[0])
    )

    html = gt_table._render_as_html()

    # Check that the footnote mark appears
    assert re.search(r"10<span[^>]*>1</span>", html)
    # Check that the text object content should appear in the footer
    assert "Bold text" in html


def test_tab_footnote_hidden_columns():
    df = pl.DataFrame(
        {
            "col1": [10],
            "col2": [100],  # Will be hidden
            "col3": [1000],
            "col4": [10000],  # Will be hidden
        }
    )

    gt_table = (
        GT(df)
        .tab_footnote(footnote="Note A", locations=loc.column_labels(columns="col1"))  # Visible
        .tab_footnote(
            footnote="Note A", locations=loc.column_labels(columns="col2")
        )  # Hidden (same text)
        .tab_footnote(
            footnote="Note A", locations=loc.column_labels(columns="col3")
        )  # Visible (same text)
        .tab_footnote(footnote="Note B", locations=loc.column_labels(columns="col2"))  # Hidden only
        .tab_footnote(
            footnote="Note B", locations=loc.column_labels(columns="col4")
        )  # Hidden only (same text)
        .tab_footnote(footnote="Note C", locations=loc.column_labels(columns="col1"))  # Visible
        .cols_hide(columns=["col2", "col4"])
    )

    html = gt_table._render_as_html()

    # Extract footnote marks from visible column headers
    col1_match = re.search(r'id="col1"[^>]*>([^<]*(?:<[^>]*>[^<]*)*)</th>', html)
    col3_match = re.search(r'id="col3"[^>]*>([^<]*(?:<[^>]*>[^<]*)*)</th>', html)

    assert col1_match is not None
    assert col3_match is not None

    col1_marks_match = re.search(
        r'<span class="gt_footnote_marks"[^>]*>([^<]*)</span>', col1_match.group(1)
    )
    col3_marks_match = re.search(
        r'<span class="gt_footnote_marks"[^>]*>([^<]*)</span>', col3_match.group(1)
    )

    # col1 should have marks 1,2 (Note A and Note C)
    assert col1_marks_match is not None
    assert col1_marks_match.group(1) == "1,2"

    # col3 should have mark 1 only (Note A)
    assert col3_marks_match is not None
    assert col3_marks_match.group(1) == "1"

    # Extract footer footnotes
    footer_matches = re.findall(
        r'<span class="gt_footnote_marks"[^>]*>([^<]*)</span>\s*([^<]+?)(?=</td>)', html
    )

    # Should only show 2 footnotes in footer (Note A and Note C)
    # Note B should not appear because it only targets hidden columns
    assert len(footer_matches) == 2

    # Check footnote texts and marks
    footnote_dict = {mark.rstrip("."): text.strip() for mark, text in footer_matches}
    assert footnote_dict["1"] == "Note A"  # Appears on visible columns
    assert footnote_dict["2"] == "Note C"  # Appears on visible column
    assert "Note B" not in html  # Should not appear anywhere since only targets hidden columns

    # Verify that duplicate footnote text gets same mark number
    # Note A appears on both col1 and col3 but should use the same mark (1)


def test_tab_footnote_mixed_locations_hidden():
    df = pl.DataFrame({"visible_col": [10], "hidden_col": [100]})

    gt_table = (
        GT(df)
        .tab_footnote(
            footnote="Mixed location note",
            locations=[
                loc.column_labels(columns="visible_col"),
                loc.column_labels(columns="hidden_col"),
            ],
        )
        .cols_hide(columns="hidden_col")
    )

    html = gt_table._render_as_html()

    # Footnote should appear because it targets at least one visible location
    assert "Mixed location note" in html

    # Mark should appear on visible column
    visible_match = re.search(r'id="visible_col"[^>]*>([^<]*(?:<[^>]*>[^<]*)*)</th>', html)
    assert visible_match is not None

    marks_match = re.search(
        r'<span class="gt_footnote_marks"[^>]*>([^<]*)</span>', visible_match.group(1)
    )
    assert marks_match is not None
    assert marks_match.group(1) == "1"
