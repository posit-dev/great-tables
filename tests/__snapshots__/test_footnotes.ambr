# serializer version: 1
# name: test_tab_footnote_complete_ordering_snapshot
  '''
  <div id="test_complete_footnote_ordering" style="padding-left:0px;padding-right:0px;padding-top:10px;padding-bottom:10px;overflow-x:auto;overflow-y:auto;width:auto;height:auto;">
  
  <table class="gt_table" data-quarto-disable-processing="false" data-quarto-bootstrap="false">
  <thead>
  
    <tr class="gt_heading">
      <td colspan="3" class="gt_heading gt_title gt_font_normal">Title<span class="gt_footnote_marks" style="white-space:nowrap;font-style:italic;font-weight:normal;line-height:0;">1</span></td>
    </tr>
    <tr class="gt_heading">
      <td colspan="3" class="gt_heading gt_subtitle gt_font_normal gt_bottom_border">Subtitle<span class="gt_footnote_marks" style="white-space:nowrap;font-style:italic;font-weight:normal;line-height:0;">2</span></td>
    </tr>
  <tr class="gt_col_headings gt_spanner_row">
    <th class="gt_col_heading gt_columns_bottom_border gt_left" rowspan="2" colspan="1" scope="col" id="test_complete_footnote_ordering-Stubhead">Stubhead<span class="gt_footnote_marks" style="white-space:nowrap;font-style:italic;font-weight:normal;line-height:0;">3</span></th>
    <th class="gt_center gt_columns_top_border gt_column_spanner_outer" rowspan="1" colspan="1" scope="col" id="test_complete_footnote_ordering-Spanner-A">
      <span class="gt_column_spanner">Spanner A<span class="gt_footnote_marks" style="white-space:nowrap;font-style:italic;font-weight:normal;line-height:0;">4</span></span>
    </th>
    <th class="gt_center gt_columns_top_border gt_column_spanner_outer" rowspan="1" colspan="1" scope="col" id="test_complete_footnote_ordering-Spanner-B">
      <span class="gt_column_spanner">Spanner B<span class="gt_footnote_marks" style="white-space:nowrap;font-style:italic;font-weight:normal;line-height:0;">5</span></span>
    </th>
  </tr>
  <tr class="gt_col_headings">
    <th class="gt_col_heading gt_columns_bottom_border gt_right" rowspan="1" colspan="1" scope="col" id="test_complete_footnote_ordering-col1">col1<span class="gt_footnote_marks" style="white-space:nowrap;font-style:italic;font-weight:normal;line-height:0;">6</span></th>
    <th class="gt_col_heading gt_columns_bottom_border gt_right" rowspan="1" colspan="1" scope="col" id="test_complete_footnote_ordering-col2">col2<span class="gt_footnote_marks" style="white-space:nowrap;font-style:italic;font-weight:normal;line-height:0;">7</span></th>
  </tr>
  </thead>
  <tbody class="gt_table_body">
    <tr>
      <th class="gt_row gt_left gt_stub">Row1<span class="gt_footnote_marks" style="white-space:nowrap;font-style:italic;font-weight:normal;line-height:0;">8</span></th>
      <td class="gt_row gt_right"><span class="gt_footnote_marks" style="white-space:nowrap;font-style:italic;font-weight:normal;line-height:0;">9</span> 10</td>
      <td class="gt_row gt_right">20</td>
    </tr>
  </tbody>
  <tfoot><tr class="gt_footnotes"><td class="gt_footnote" colspan="3"><span class="gt_footnote_marks" style="white-space:nowrap;font-style:italic;font-weight:normal;line-height:0;">1</span> Title note</td></tr><tr class="gt_footnotes"><td class="gt_footnote" colspan="3"><span class="gt_footnote_marks" style="white-space:nowrap;font-style:italic;font-weight:normal;line-height:0;">2</span> Subtitle note</td></tr><tr class="gt_footnotes"><td class="gt_footnote" colspan="3"><span class="gt_footnote_marks" style="white-space:nowrap;font-style:italic;font-weight:normal;line-height:0;">3</span> Stubhead note</td></tr><tr class="gt_footnotes"><td class="gt_footnote" colspan="3"><span class="gt_footnote_marks" style="white-space:nowrap;font-style:italic;font-weight:normal;line-height:0;">4</span> Spanner A note</td></tr><tr class="gt_footnotes"><td class="gt_footnote" colspan="3"><span class="gt_footnote_marks" style="white-space:nowrap;font-style:italic;font-weight:normal;line-height:0;">5</span> Spanner B note</td></tr><tr class="gt_footnotes"><td class="gt_footnote" colspan="3"><span class="gt_footnote_marks" style="white-space:nowrap;font-style:italic;font-weight:normal;line-height:0;">6</span> Col1 note</td></tr><tr class="gt_footnotes"><td class="gt_footnote" colspan="3"><span class="gt_footnote_marks" style="white-space:nowrap;font-style:italic;font-weight:normal;line-height:0;">7</span> Col2 note</td></tr><tr class="gt_footnotes"><td class="gt_footnote" colspan="3"><span class="gt_footnote_marks" style="white-space:nowrap;font-style:italic;font-weight:normal;line-height:0;">8</span> Stub note</td></tr><tr class="gt_footnotes"><td class="gt_footnote" colspan="3"><span class="gt_footnote_marks" style="white-space:nowrap;font-style:italic;font-weight:normal;line-height:0;">9</span> Body note</td></tr></tfoot>
  </table>
  
  </div>
          
  '''
# ---
# name: test_tab_footnote_stub_body_ordering_snapshot
  '''
  <tbody class="gt_table_body">
    <tr>
      <th class="gt_row gt_left gt_stub">A<span class="gt_footnote_marks" style="white-space:nowrap;font-style:italic;font-weight:normal;line-height:0;">1</span></th>
      <td class="gt_row gt_left">Y</td>
    </tr>
    <tr>
      <th class="gt_row gt_left gt_stub">B</th>
      <td class="gt_row gt_left">Z<span class="gt_footnote_marks" style="white-space:nowrap;font-style:italic;font-weight:normal;line-height:0;">2</span></td>
    </tr>
  </tbody>
  '''
# ---
